README.md
# aoc2022
Advent of code solutions in Clojure


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2022-12-03
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2022-12-03
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/aoc2022/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/aoc2022/compare/0.1.0...0.1.1


test/aoc2022/core_test.clj
(ns aoc2022.core-test
  (:require [clojure.test :refer :all]
            [aoc2022.util :refer :all]
            [aoc2022.week1 :refer :all]
            [aoc2022.week2 :refer :all]))

(deftest d01-sample-test
  (testing "Day 1 Task 1 sample"
    (is (= 24000 (solve-d01-1 (input "dec01sample.txt")))))
  (testing "Day 1 Task 2 sample"
    (is (= 45000 (solve-d01-2 (input "dec01sample.txt"))))))

(deftest d02-sample-test
  (testing "Day 2 Task 1 sample"
    (is (= 15 (solve-d02-1 (input "dec02sample.txt")))))
  (testing "Day 2 Task 2 sample"
    (is (= 12 (solve-d02-2 (input "dec02sample.txt"))))))

(deftest d03-sample-test
  (testing "Day 3 Task 1 sample"
    (is (= 157 (solve-d03-1 (input "dec03sample.txt")))))
  (testing "Day 3 Task 2 sample"
    (is (= 70 (solve-d03-2 (input "dec03sample.txt"))))))

(deftest d04-sample-test
  (testing "Day 4 Task 1 sample"
    (is (= 2 (solve-d04-1 (input "dec04sample.txt")))))
  (testing "Day 4 Task 2 sample"
    (is (= 4 (solve-d04-2 (input "dec04sample.txt"))))))

(deftest d06-sample-test
  (testing "Day 4 Task 1 sample"
    (is (= 2 (day6-task1 (parse "dec06sample.txt"))))))



src/aoc2022/core.clj
(ns aoc2022.core
  (:gen-class)
  (:require [clojure.java.io :as io]
            [clojure.string :as str]))

(defn -main []
  (println "Solving Advent of Code 2022"))

(defn input
  "Read a file from resource folder"
  [file]
  (slurp (io/resource file)))

(defn split-by
  "Reverse split arguments for easier threading"
  [regex text]
  (str/split text regex))

(defn d01-add-strings
  "Add numbers in a list of strings"
  [strings]
  (reduce #(+ %1 (Integer/parseInt %2)) 0 strings))

(defn d01-bags
  "Parse the data and summarize each bag"
  [data]
  (let [groups (str/split data #"\n\n")
        elves (map #(str/split % #"\n") groups)]
    (map d01-add-strings elves)))

(defn solve-d01-1
  "Day 1 Task 1"
  [data]
  (->> data
       (d01-bags)
       (apply max)))

(defn solve-d01-2
  "Day 1 Task 2"
  [data]
  (->> data
       (d01-bags)
       (sort)
       (reverse)
       (take 3)
       (reduce +)))

; Day 1 completed

(def d02-move-value {:rock 1 :paper 2 :scissors 3})

(defn d02-match-value [[p1 p2]]
  (cond
    (= p1 p2) 3
    (or (= [p1 p2] [:rock :scissors])
        (= [p1 p2] [:paper :rock])
        (= [p1 p2] [:scissors :paper])) 6
    :else 0))


(defn d02-parse-move [[m1 _ m2]]
  (let [meaning {\A :rock \B :paper \C :scissors
                 \X :rock \Y :paper \Z :scissors}]
    (map meaning [m1 m2])))

(defn d02-parse-expectation [[m1 _ m2]]
  (let [meaning {\A :rock \B :paper \C :scissors
                 \X :lose \Y :draw \Z :win}]
    (map meaning [m1 m2])))

(defn d02-determine-move [pattern]
  (let [path {:rock {:win :paper :draw :rock :lose :scissors}
              :paper {:win :scissors :draw :paper :lose :rock}
              :scissors {:win :rock :draw :scissors :lose :paper}}]
    (get-in path pattern)))

(defn d02-p2-score [[p1 p2]]
  (+ (d02-move-value p2) (d02-match-value [p2 p1])))

(defn solve-d02-1
  "Day 2 Task 1"
  [data]
  (let [moves (map d02-parse-move (str/split data #"\n"))]
    (reduce + (map d02-p2-score moves))))

(defn solve-d02-2
  "Day 2 Task 2"
  [data]
  (->> data
       (split-by #"\n")
       (map d02-parse-expectation)
       (map (fn [[enemy expected]]
              [enemy (d02-determine-move [enemy expected])]))
       (map d02-p2-score)
       (reduce +)))

; Day 2 completed

(defn d03-compartmentalize [bag]
  (split-at (/ (count bag) 2) bag))

(defn d03-find-common [groups]
  (first (apply clojure.set/intersection (map set groups))))

(defn d03-item-value [item]
  (let [offset (if (Character/isLowerCase item) 96 38)]
    (- (int item) offset)))

(defn solve-d03-1
  "Day 3 Task 1"
  [data]
  (->> data
       (split-by #"\n")
       (map d03-compartmentalize)
       (map d03-find-common)
       (map d03-item-value)
       (reduce +)))

(defn solve-d03-2
  "Day 3 Task 2"
  [data]
  (->> data
       (split-by #"\n")
       (partition 3)
       (map d03-find-common)
       (map d03-item-value)
       (reduce +)))

; Day 3 completed

(defn d04-numbers [s]
  (let [pair (split-by #"-" s)
        [start end] (map read-string pair)]
    (set (range start (inc end)))))

(defn d04-ranges [[r1 r2]]
  [(d04-numbers r1) (d04-numbers r2)])

(defn d04-covered? [[r1 r2]]
  (or (every? r1 r2) (every? r2 r1)))

(defn solve-d04-1
  "Day 4 Task 1"
  [data]
  (->> data
       (split-by #"\n")
       (map (partial split-by #","))
       (map d04-ranges)
       (filter d04-covered?)
       count))

(defn solve-d04-2
  "Day 4 Task 2"
  [data]
  (->> data
       (split-by #"\n")
       (map (partial split-by #","))
       (map d04-ranges)
       (filter (fn [[r1 r2]] (some r1 r2)))
       count))

(comment
  (solve-d01-1 (input "dec01.txt"))
  (solve-d01-2 (input "dec01.txt"))
  (solve-d02-1 (input "dec02.txt"))
  (solve-d02-2 (input "dec02.txt"))
  (solve-d03-1 (input "dec03.txt"))
  (solve-d03-2 (input "dec03.txt"))
  (solve-d04-1 (input "dec04.txt"))
  (solve-d04-2 (input "dec04.txt")))


src/aoc2022/week2.clj
(ns aoc2022.week2
  (:gen-class)
  (:require [clojure.string :as str])
  (:require [aoc2022.util :refer [parse]]))

; Day 5

(defn str->move [raw-move]
  (let [[times from to]
        (->> raw-move (re-seq #"\d+") (map #(Integer/valueOf %)))]
    (repeat times [from to])))

(defn str->bucket [raw-stack]
  (->> raw-stack
       (concat " ")
       (partition 4)
       (map #(nth % 2))
       (apply str)))

(defn day5-parse [lines]
  (let [[raw-stacks _ raw-moves] (partition-by str/blank? lines)
        stacks (->> raw-stacks
                    (drop-last)
                    (map str->bucket)
                    (apply mapv vector)
                    (map (fn [c] (filter #(not= \space %) c)))
                    (zipmap (map inc (range)))
                    (into (sorted-map)))
        moves (->> raw-moves
                   (map str->move)
                   (apply conj)
                   (flatten)
                   (partition 2))]
    [stacks moves]))

(defn move-stack [stack [from to]]
  (-> stack
      (update-in [to] #(cons (first (get-in stack [from])) %))
      (update-in [from] #(drop 1 %))))

(defn day5-task1 [data]
  (->>
   (loop [[stacks moves] (day5-parse data)]
     (if (empty? moves) stacks
         (recur [(move-stack stacks (last moves))
                 (butlast moves)])))
   vals
   (map first)
   (apply str)))

(comment
  (day5-task1 (parse "dec05.txt")))

; TODO DAY 5 TASK 2

; Day 6

(defn find-marker [n msg]
  (->> (partition n 1 msg)
       (map-indexed (fn [i s] [(+ n i) (set s)]))
       (filter #(= n (count (second %))))
       first first))

(defn day6-task1 [data]
  (->> (first data) (find-marker 4)))

(defn day6-task2 [data]
  (->> (first data) (find-marker 14)))

(comment
  (day6-task1 (parse "dec06.txt"))
  (day6-task2 (parse "dec06.txt")))

; Day 7

(defn day7-task1 [data]
  (loop [cmds (drop 1 data) cwd ["/"] dir {"/" {}}]
    (if (empty? cmds) dir
        (let [cmd (first cmds)
              [w1 w2 & ws] (str/split cmd #" ")
              newcwd (cond
                   ;(every? #(Character/isDigit %) w1)
                       (= cmd "$ cd ..") (drop-last cwd)
                       (= [w1 w2] ["$" "cd"]) (into [] (concat cwd ws))
                       :else cwd)
              newdir (cond
                       (= w1 "dir")
                       (assoc-in dir cwd
                                 (into (get-in dir cwd)
                                       {w2 {}}))
                       (every? #(Character/isDigit %) w1)
                       (assoc-in dir cwd
                                 (into (get-in dir cwd)
                                       {w2 (Integer/parseInt w1)}))
                       :else dir)]
          (recur (drop 1 cmds) newcwd newdir)))))

; TODO to finish


(comment
  (clojure.pprint/pprint (day7-task1 (parse "dec07sample.txt"))))

; the lines must be traversed keeping cwd in state
; directory structure represented as nested maps?
; key is the name of file or directory
; in case of dir, value is another map
; in case of file, value is the size
; state = ["/" "a"]
; {"\" {"a" {}}}

src/aoc2022/week1.clj
(ns aoc2022.week1
  (:gen-class)
  (:require [clojure.java.io :as io]
            [clojure.string :as str]))

(defn -main []
  (println "Solving Advent of Code 2022"))

(defn input
  "Read a file from resource folder"
  [file]
  (slurp (io/resource file)))

(defn split-by
  "Reverse split arguments for easier threading"
  [regex text]
  (str/split text regex))

(defn d01-add-strings
  "Add numbers in a list of strings"
  [strings]
  (reduce #(+ %1 (Integer/parseInt %2)) 0 strings))

(defn d01-bags
  "Parse the data and summarize each bag"
  [data]
  (let [groups (str/split data #"\n\n")
        elves (map #(str/split % #"\n") groups)]
    (map d01-add-strings elves)))

(defn solve-d01-1
  "Day 1 Task 1"
  [data]
  (->> data
       (d01-bags)
       (apply max)))

(defn solve-d01-2
  "Day 1 Task 2"
  [data]
  (->> data
       (d01-bags)
       (sort)
       (reverse)
       (take 3)
       (reduce +)))

; Day 1 completed

(def d02-move-value {:rock 1 :paper 2 :scissors 3})

(defn d02-match-value [[p1 p2]]
  (cond
    (= p1 p2) 3
    (or (= [p1 p2] [:rock :scissors])
        (= [p1 p2] [:paper :rock])
        (= [p1 p2] [:scissors :paper])) 6
    :else 0))


(defn d02-parse-move [[m1 _ m2]]
  (let [meaning {\A :rock \B :paper \C :scissors
                 \X :rock \Y :paper \Z :scissors}]
    (map meaning [m1 m2])))

(defn d02-parse-expectation [[m1 _ m2]]
  (let [meaning {\A :rock \B :paper \C :scissors
                 \X :lose \Y :draw \Z :win}]
    (map meaning [m1 m2])))

(defn d02-determine-move [pattern]
  (let [path {:rock {:win :paper :draw :rock :lose :scissors}
              :paper {:win :scissors :draw :paper :lose :rock}
              :scissors {:win :rock :draw :scissors :lose :paper}}]
    (get-in path pattern)))

(defn d02-p2-score [[p1 p2]]
  (+ (d02-move-value p2) (d02-match-value [p2 p1])))

(defn solve-d02-1
  "Day 2 Task 1"
  [data]
  (let [moves (map d02-parse-move (str/split data #"\n"))]
    (reduce + (map d02-p2-score moves))))

(defn solve-d02-2
  "Day 2 Task 2"
  [data]
  (->> data
       (split-by #"\n")
       (map d02-parse-expectation)
       (map (fn [[enemy expected]]
              [enemy (d02-determine-move [enemy expected])]))
       (map d02-p2-score)
       (reduce +)))

; Day 2 completed

(defn d03-compartmentalize [bag]
  (split-at (/ (count bag) 2) bag))

(defn d03-find-common [groups]
  (first (apply clojure.set/intersection (map set groups))))

(defn d03-item-value [item]
  (let [offset (if (Character/isLowerCase item) 96 38)]
    (- (int item) offset)))

(defn solve-d03-1
  "Day 3 Task 1"
  [data]
  (->> data
       (split-by #"\n")
       (map d03-compartmentalize)
       (map d03-find-common)
       (map d03-item-value)
       (reduce +)))

(defn solve-d03-2
  "Day 3 Task 2"
  [data]
  (->> data
       (split-by #"\n")
       (partition 3)
       (map d03-find-common)
       (map d03-item-value)
       (reduce +)))

; Day 3 completed

(defn d04-numbers [s]
  (let [pair (split-by #"-" s)
        [start end] (map read-string pair)]
    (set (range start (inc end)))))

(defn d04-ranges [[r1 r2]]
  [(d04-numbers r1) (d04-numbers r2)])

(defn d04-covered? [[r1 r2]]
  (or (every? r1 r2) (every? r2 r1)))

(defn solve-d04-1
  "Day 4 Task 1"
  [data]
  (->> data
       (split-by #"\n")
       (map (partial split-by #","))
       (map d04-ranges)
       (filter d04-covered?)
       count))

(defn solve-d04-2
  "Day 4 Task 2"
  [data]
  (->> data
       (split-by #"\n")
       (map (partial split-by #","))
       (map d04-ranges)
       (filter (fn [[r1 r2]] (some r1 r2)))
       count))

(comment
  (solve-d01-1 (input "dec01.txt"))
  (solve-d01-2 (input "dec01.txt"))
  (solve-d02-1 (input "dec02.txt"))
  (solve-d02-2 (input "dec02.txt"))
  (solve-d03-1 (input "dec03.txt"))
  (solve-d03-2 (input "dec03.txt"))
  (solve-d04-1 (input "dec04.txt"))
  (solve-d04-2 (input "dec04.txt")))
